Font Library REST API endpoints: address initial feedback from feature branch

creativecoder opened the pull request:
## What?

Addresses the initial feedback in https://github.com/WordPress/gutenberg/pull/57688, including

- Registering `wp_font_face` with `show_in_rest` as `true` and removing the manual route registration
- Removing the constructors of both `WP_REST_Font_Families_Controller` and `WP_REST_Font_Faces_Controller`, and depending on the parent constructor
- Removing the `register_routes` method of `WP_REST_Font_Families_Controller` and depending on the parent method
- Replacing custom `permission_callback` methods and overriding the default ones, where needed
- Adding a context param to both `WP_REST_Font_Families_Controller` and `WP_REST_Font_Faces_Controller` that defaults to `edit`
- Adding tests for context param
- Conditionally adding all fields in `prepare_item_for_response` to support the `_fields` arg
- Replace `get_posts` calls with `WP_Query` and turning off the meta and term caches
- Adding translation functions to strings where they were missing
- Unsetting unwanted collection params from `parent::get_collection_params()`

Also, some bug fixes and improvements

- Turning of batching for both `WP_REST_Font_Families_Controller` and `WP_REST_Font_Faces_Controller`, since it doesn't work with file uploads
- Prevent font faces from being created or deleted when the parent font family id in the route is invalid
- Fix race condition that sometimes causes unit tests to fail
- Renaming hooked functions for deleting font family and font face posts to make it clearer what they're for
- Adding a test to make sure embedding works correctly

## Why?

See https://github.com/WordPress/gutenberg/issues/55278

## Testing Instructions

Minimal functionality changes

- Review code changes against feedback comments
- Smoke test font library
- Ensure unit tests are still passing
- Try creating or deleting a font face using an invalid font family id in the url: you should now get an error

creativecoder committed:
Addresses initial feedback on font controllers

jffng reviewed (commented):
Giving this a smoke test, and I'm getting a 404 in the embedded font_faces in the response from the `/font-families` endpoint:

```json
 "_embedded": {
            "font_faces": [
                {
                    "code": "rest_font_face_parent_id_mismatch",
                    "message": "The font face does not belong to the specified font family with id of \"147\"",
                    "data": {
                        "status": 404
                    }
                },
                ...
```

I can confirm the parent font family post has an ID of 147, and it works as expected on the feature branch. Looking into why this breaks on the PR.

EDIT

In [this comparison](https://github.com/WordPress/gutenberg/blob/e59a08043f6cec11c3afde26e07a9c7b63ad6f24/lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php#L269), `(int) $response->data['parent']` is evaluating to `0` in my testing, so the error condition is met and thrown.

creativecoder committed:
Remove font face controller constructor and a few other tweaks

creativecoder committed:
Completes implementing context param so it works correctly with embedding font faces

creativecoder committed:
Prevent creating or deleting font faces with invalid font families

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-families-controller.php:
```diff
@@ -254,15 +198,7 @@ public function delete_item( $request ) {
 			);
 		}
 
-		$deleted = parent::delete_item( $request );
-
-		if ( is_wp_error( $deleted ) ) {
-			return $deleted;
-		}
-
-		foreach ( $this->get_font_face_ids( $font_family_id ) as $font_face_id ) {
-			wp_delete_post( $font_face_id, true );
-		}
```
Deleting child font families is now handled by callbacks hooked into `deleted_post`, so we don't need to do it here in the controller

creativecoder committed:
Fix indenting and clarify comment

creativecoder commented:
> Giving this a smoke test, and I'm getting a 404 in the embedded font_faces in the response from the /font-families endpoint:

Thank you @jffng for catching this. I've fixed the issue, and added a unit test to catch any regressions in the future.

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-autosave-fonts-controller.php:
```diff
@@ -0,0 +1,25 @@
+<?php
+/**
+ * WP_REST_Autosave_Fonts_Controller class.
+ *
+ * This file contains the class for the REST API Autosave Font Families and Font Faces Controller.
+ *
+ * @package    WordPress
+ * @subpackage REST_API
+ * @since      6.5.0
+ */
+
+if ( class_exists( 'WP_REST_Autosave_Fonts_Controller' ) ) {
+	return;
+}
+
+/**
+ * Autosave Font Families Controller class.
+ *
+ * @since 6.5.0
+ */
+class WP_REST_Autosave_Fonts_Controller {
+	public function register_routes() {
+		// disable autosave endpoints for font families and faces.
+	}
+}
```
Here I'm making the empty autosave controller generic so we can use it for both `wp_font_family` and `wp_font_face` post types.

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php:
```diff
@@ -55,7 +33,7 @@ public function __construct() {
 		register_rest_route(
 			$this->namespace,
-			'/' . $this->parent_base . '/(?P<font_family_id>[\d]+)/' . $this->rest_base,
+			'/' . $this->rest_base,
```
By defining the `rest_base` as `font-families/(?P<font_family_id>[\d]+)/font-faces`, we can use this `rest_base` directly for registering the routes.

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php:
```diff
@@ -16,34 +16,12 @@
-		$this->parent_base      = $parent_post_type_obj->rest_base;
-		$this->namespace        = $parent_post_type_obj->rest_namespace;
-	}
+	protected $allow_batch = false;
```
Batch support is on by default. Here I'm turning it off for font families, because the batch endpoint doesn't really support file uploads.

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php:
```diff
@@ -16,34 +16,12 @@
-	 * @var string
+	 * @var false
 	 */
-	private $parent_base;
-
-	/**
-	 * Parent font family post type
-	 *
-	 * @since 6.5.0
-	 * @var string
-	 */
-	private $parent_post_type;
-
-	public function __construct() {
-		$post_type       = 'wp_font_face';
-		$this->post_type = $post_type;
-
-		$post_type_obj   = get_post_type_object( $post_type );
-		$this->rest_base = $post_type_obj->rest_base;
-
-		$parent_post_type       = 'wp_font_family';
-		$this->parent_post_type = $parent_post_type;
-		$parent_post_type_obj   = get_post_type_object( $parent_post_type );
-		$this->parent_base      = $parent_post_type_obj->rest_base;
-		$this->namespace        = $parent_post_type_obj->rest_namespace;
-	}
```
Removing the constructor, as requested here: https://github.com/WordPress/gutenberg/pull/57688#discussion_r1456271827

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php:
```diff
@@ -55,7 +33,7 @@ public function __construct() {
 	public function register_routes() {
```
Still overriding `register_routes` for font faces, as we don't want the `WP_REST_Server::EDITABLE` route that's created by default for updating existing font faces (which could get quite complex with file uploads)

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php:
```diff
@@ -67,7 +45,7 @@ public function register_routes() {
 					'methods'             => WP_REST_Server::READABLE,
 					'callback'            => array( $this, 'get_items' ),
-					'permission_callback' => array( $this, 'get_font_faces_permissions_check' ),
+					'permission_callback' => array( $this, 'get_items_permissions_check' ),
```
Overriding the default permissions check methods, as requested here: https://github.com/WordPress/gutenberg/pull/57688#discussion_r1456276097

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php:
```diff
@@ -99,8 +77,10 @@ public function register_routes() {
 					'callback'            => array( $this, 'get_item' ),
-					'permission_callback' => array( $this, 'get_font_faces_permissions_check' ),
-					'args'                => array(),
+					'permission_callback' => array( $this, 'get_item_permissions_check' ),
+					'args'                => array(
+						'context' => $this->get_context_param( array( 'default' => 'edit' ) ),
+					),
```
Overriding default permission method and adding `context` param to route.

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php:
```diff
@@ -124,12 +104,13 @@ public function register_routes() {
 
-		if ( ! current_user_can( $post_type->cap->edit_posts ) ) {
+		if ( ! current_user_can( $post_type->cap->read ) ) {
 			return new WP_Error(
```
Here checking `cap->read` rather than `cap->edit_posts`. In reality they all map to `edit_theme_options`, but checking for read capability seems more correct when getting items.

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php:
```diff
@@ -140,6 +121,18 @@ public function get_font_faces_permissions_check() {
+	public function get_item_permissions_check( $request ) {
+		return $this->get_items_permissions_check( $request );
+	}
+
```
Overriding the default permissions method for `get_item`

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php:
```diff
@@ -263,14 +256,12 @@ public function get_item( $request ) {
-
-		if ( (int) $font_family->ID !== (int) $response->data['parent'] ) {
+		if ( (int) $font_family->ID !== (int) $post->post_parent ) {
 			return new WP_Error(
```
We need to check the `$post->post_parent` directly here, since `$response->data['parent']` may not be in the response if the `_fields` param is used.

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php:
```diff
@@ -291,19 +282,26 @@ public function get_item( $request ) {
+		if ( is_wp_error( $font_family ) ) {
+			return $font_family;
+		}
+
```
Checking that the parent font family exists before creating a new font face.

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php:
```diff
@@ -291,19 +282,26 @@ public function get_item( $request ) {
 		);
-		if ( ! empty( $existing_font_face ) ) {
+		if ( ! empty( $query->get_posts() ) ) {
 			return new WP_Error(
```
Using WP_Query with meta/term caches turned off, rather than get_posts, as requested.

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php:
```diff
@@ -367,9 +365,28 @@ public function create_item( $request ) {
+			);
+		}
+
 		$force = isset( $request['force'] ) ? (bool) $request['force'] : false;
```
Check that the parent font family exists and belongs to the font face before deleting the font face.

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php:
```diff
@@ -391,19 +408,46 @@ public function delete_item( $request ) {
+		 * @param WP_Post          $post     Font face post object.
+		 * @param WP_REST_Request  $request  Request object.
+		 */
+		return apply_filters( 'rest_prepare_wp_font_face', $response, $item, $request );
```
Adding each response property conditionally, so it can be toggled with the `_fields` param.

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php:
```diff
@@ -427,6 +471,7 @@ public function get_item_schema() {
 				'id'                 => array(
 					'description' => __( 'Unique identifier for the post.', 'default' ),
 					'type'        => 'integer',
+					'context'     => array( 'edit', 'embed' ),
```
The context settings here allows each property to show in the edit context (the default) and show when embedded when requesting the parent font family.

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php:
```diff
@@ -551,13 +599,31 @@ public function get_item_schema() {
+		 *
+		 * @param array $query_params JSON Schema-formatted collection parameters.
+		 */
+		return apply_filters( 'rest_wp_font_face_collection_params', $query_params );
```
Unsetting unneeded params, as requested.

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php:
```diff
@@ -583,18 +649,6 @@ public function get_create_params() {
-	protected function check_is_post_type_allowed( $post_type ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable -- required by parent class
-		return true;
-	}
-
```
This is not needed now that we have `show_in_rest` set to `true` for `wp_font_face`.

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php:
```diff
@@ -603,7 +657,7 @@ protected function check_is_post_type_allowed( $post_type ) { // phpcs:ignore Va
 	 * @return WP_Post|WP_Error Post object if ID is valid, WP_Error otherwise.
 	 */
-	protected function get_font_family_post( $font_family_id ) {
+	protected function get_parent_font_family_post( $font_family_id ) {
```
Renaming this for clarity that it's getting the parent post.

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php:
```diff
@@ -637,13 +691,13 @@ protected function prepare_links( $post ) {
 			'parent'     => array(
-				'href' => rest_url( $this->namespace . '/' . $this->parent_base . '/' . $post->post_parent ),
+				'href' => rest_url( $this->namespace . '/font-families/' . $post->post_parent ),
 			),
```
parent_base is no longer being set by the constructor, so manually using `font-families` string for the links here.

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-families-controller.php:
```diff
@@ -18,81 +18,12 @@
-				),
-			)
-		);
-	}
```
Removing the constructor and register_routes methods to depend on the parents, as requested.

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-families-controller.php:
```diff
@@ -18,81 +18,12 @@
-			)
-		);
-	}
+	protected $allow_batch = false;
```
Setting batch to false here, as well, as we want to leave the possibility of adding image uploads for previews to the controller in the future.

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-families-controller.php:
```diff
@@ -102,20 +33,32 @@ public function register_routes() {
+	public function get_item_permissions_check( $request ) {
+		return $this->get_items_permissions_check( $request );
+	}
+
```
Overriding the default permission callbacks from the parent controller.

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-families-controller.php:
```diff
@@ -276,19 +210,47 @@ public function delete_item( $request ) {
+		 * @param WP_Post          $post     Font family post object.
+		 * @param WP_REST_Request  $request  Request object.
+		 */
+		return apply_filters( 'rest_prepare_wp_font_family', $response, $item, $request );
```
Support the `_fields` param.

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-families-controller.php:
```diff
@@ -312,6 +274,7 @@ public function get_item_schema() {
 				'id'                   => array(
 					'description' => __( 'Unique identifier for the post.', 'default' ),
 					'type'        => 'integer',
+					'context'     => array( 'edit' ),
```
Adding support for the context param. Font families are not embedded, so the only context is `edit`.

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-families-controller.php:
```diff
@@ -370,58 +336,72 @@ public function get_item_schema() {
+		 *
+		 * @param array $query_params JSON Schema-formatted collection parameters.
+		 */
+		return apply_filters( 'rest_wp_font_family_collection_params', $query_params );
```
Unsetting collection params from the parent controller, as requested.

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-families-controller.php:
```diff
@@ -370,58 +336,72 @@ public function get_item_schema() {
-
-		return true;
+		return parent::get_context_param( $args );
 	}
```
This sets the context param default to `edit`, since we don't need a `view` context for fonts as admin only settings.

creativecoder commented on lib/experimental/fonts/font-library/class-wp-rest-font-families-controller.php:
```diff
@@ -370,58 +336,72 @@ public function get_item_schema() {
+			);
+		}
+
+		return parent::get_endpoint_args_for_item_schema( $method );
```
This overrides the default method for getting the allowed params when creating or editing a font face.

creativecoder commented on lib/experimental/fonts/font-library/font-library.php:
```diff
@@ -45,26 +45,25 @@ function gutenberg_init_font_library_routes() {
-		'autosave_rest_controller_class' => 'WP_REST_Autosave_Font_Families_Controller',
-		'query_var'                      => false,
+		'autosave_rest_controller_class' => 'WP_REST_Autosave_Fonts_Controller',
 	);
```
Ordering the post type settings for `wp_font_family` so they match with `wp_font_face`.

creativecoder commented on lib/experimental/fonts/font-library/font-library.php:
```diff
@@ -79,17 +78,18 @@ function gutenberg_init_font_library_routes() {
-			'query_var'    => false,
+			'map_meta_cap'                   => false,
+			'query_var'                      => false,
+			'show_in_rest'                   => true,
+			'rest_base'                      => 'font-families/(?P<font_family_id>[\d]+)/font-faces',
+			'rest_controller_class'          => 'WP_REST_Font_Faces_Controller',
+			'autosave_rest_controller_class' => 'WP_REST_Autosave_Fonts_Controller',
```
These are the new post registration settings here, using the built in plumbing for registering the routes for `wp_font_face`.

creativecoder commented on lib/experimental/fonts/font-library/font-library.php:
```diff
@@ -242,5 +242,5 @@ function _wp_delete_font_face( $post_id, $post ) {
 		}
 	}
-	add_action( 'before_delete_post', '_wp_delete_font_face', 10, 2 );
+	add_action( 'before_delete_post', '_wp_before_delete_font_face', 10, 2 );
```
Renaming all of these hooked functions for clarity.

creativecoder reviewed (commented):
I've left inline comments to explain the changes in hopes of making this easier to review.

jffng reviewed (approved):
Verified the following: 

✅ Code changes address relevant feedback
✅ Smoke test font library — can still install and uninstall local and collection fonts
✅ Unit tests pass
✅ Creating and deleting a font face with invalid font family ID fails and error is thrown
